internal/yggdrasil/yggdrasil.go

```diff
@@ -416,3 +432,134 @@ func (h *Handler) setStorageConfiguration(ctx context.Context,
 
 	return err
 }
+
+func (h *Handler) readAndValidateSecret(ctx context.Context, secretName, secretNamespace string, secretKeys keyMapType) (*corev1.Secret, error) {
+	optional := secretKeys == nil
+	secretObj := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: secretNamespace}}
+	err := h.client.Get(ctx, client.ObjectKeyFromObject(secretObj), secretObj)
+	if err != nil {
+		if errors.IsNotFound(err) && optional {
+			return nil, nil
+		}
+		return nil, err
+	}
+	for key := range secretKeys {
+		if secretObj.Data != nil {
+			if _, ok := secretObj.Data[key]; ok {
+				continue
+			}
+		}
+		return nil, fmt.Errorf("missing secret key. secret: %s. key: %s. Namespace: %s", secretName, key, secretNamespace)
+	}
+	return secretObj, nil
+}
+
+func addSecretToSecretList(secretList *models.SecretList, secret *corev1.Secret) error {
+	dataMap := map[string]string{}
+	for name, value := range secret.Data {
+		dataMap[name] = strfmt.Base64(value).String()
+	}
+	dataJson, err := json.Marshal(dataMap)
+	if err != nil {
+		return err
+	}
+	*secretList = append(*secretList, &models.Secret{
+		Data: string(dataJson),
+		Name: secret.Name,
+	})
+
+	return nil
+}
+
+func (h *Handler) createSecretList(ctx context.Context, logger logr.Logger, deployments []v1alpha1.EdgeDeployment, device *v1alpha1.EdgeDevice) (models.SecretList, error) {
+	list := models.SecretList{}
+
+	// create map of secret names and keys
+	secretMap := secretMapType{}
+	for _, deployment := range deployments {
+		podSpec := deployment.Spec.Pod.Spec
+		allContainers := append(podSpec.InitContainers, podSpec.Containers...)
+		for _, container := range allContainers {
+			extractSecretsFromContainer(&container, secretMap)
+		}
+	}
+
+	// read secrets and add to secrets list
+	for name, keys := range secretMap {
+		secretObj, err := h.readAndValidateSecret(ctx, name, device.Namespace, keys)
+		if err != nil {
+			return nil, err
+		}
+		if secretObj == nil {
+			continue
+		}
+		err = addSecretToSecretList(&list, secretObj)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return list, nil
+}
+
+// return value maps a secret name to its mandatory keys. nil keys indicates optional secret
+func extractSecretsFromContainer(container *corev1.Container, secretMap secretMapType) {
+	extractSecretsFromEnvFrom(container.EnvFrom, secretMap)
+	extractSecretsFromEnv(container.Env, secretMap)
+}
+
+func extractSecretsFromEnvFrom(envFrom []corev1.EnvFromSource, secretMap secretMapType) {
+	if len(envFrom) == 0 {
```

Comment:
why this len? No need at all there is nothing to range no?

Reply:
right